torch/quantization/make_module.py

```diff
@@ -0,0 +1,61 @@
+
+import torch
+
+_func_from_float = {
+    torch.add: torch.ops.quantized.add
+}
+
+"""Wrapper class for functions.
+
+Args:
+    func: The function to wrap the module around
+    name: Name of the module class. If None, the class name will be capitalized
+          op name.
+"""
+class _WrapModule(torch.nn.Module):
```

Comment:
With this approach, type(mod) will be <class '__main__._WrapModule>. An approach (inspired from Dima's earlier suggestion is to do codegen:

```python
import torch
import torch.nn as nn

def wrapper_forward(self, *args, **kwargs):
        # Expand tuple
        return self.operation(*args, **kwargs)

def create_module(operation):
    C= type(
        'torch_ops_'+operation.__name__,
        (nn.Module,),
        {
            "forward": wrapper_forward,
            "operation": operation,
        }
    )
    # Specify the type in globals()
    globals()[C.__name__] = C
    return C   

C1 = create_module(torch.add)()
print(C1.operation)
print(C1(2,3))
print(type(C1))
if type(C1) == torch_ops_add:
    print('Yes')
```  
Prints:
<built-in method add of type object at 0x7f4336df1a60>
tensor(5)
<class '__main__.torch_ops_add'>
Yes

Reply:
cc @dzhulgakov

Reply:
There is another approach that we extensively discussed with @jerryzh168 yesterday. It achieves the same thing, and does not involce the use of globals:

```python
"""Factory method to create a module wrapper.

Args:
    func: The function to wrap the module around
    name: Name of the module class. If None, the class name will be capitalized op name.

Returns:
    Module class
"""
def make_module(op, quant_op=None, name=None):
    name = op.__name__.capitalize() if name is None else name
    
    if quant_op is None:
        quant_op = func_from_float.get(op, None)
    
    def init(self, func):
        super(type(self), self).__init__()
        self.func = func
        
    def forward(self, *args, **kwargs):
        return self.func(*args, **kwargs)
    
    def from_float(self, *args, **kwargs):
        return make_module(quant_op, name=name + "_quantized")

    module = type(name, 
                 (torch.nn.Module,),
                 {"__init__": init,
                    "forward": forward,
                    "from_float": from_float})
    return module(op)
```

This prints: 

```python
mod = make_module(torch.add)
mod_quant = mod.from_float()

a = torch.tensor(1, dtype=torch.float)
qa = torch.quantize_linear(a, 1.0, 127, torch.quint8)
b = torch.tensor(2, dtype=torch.float)
qb = torch.quantize_linear(b, 1.0, 127, torch.quint8)

print("Types\t\t:", type(mod), type(mod_quant))
print("__repr__\t:", mod, mod_quant)
print("forward call\t:", mod(a, b))
print("forward call\t:", mod_quant(qa, qb, 1.0, 0))
```

```
Types		: <class '__main__.Add'> <class '__main__.Add_quantized'>
__repr__	: Add() Add_quantized()
forward call	: tensor(3.)
forward call	: tensor(3., size=(), dtype=torch.quint8, scale=1.0, zero_point=0)
```